server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OffHeapStarTreeBuilder.java

```diff
@@ -0,0 +1,896 @@
+                }
+
+                int _docId = 1;
+
+                @Override
+                public boolean hasNext() {
+                    return _hasNext;
+                }
+
+                @Override
+                public StarTreeDocument next() {
+                    StarTreeDocument next = reduceSegmentStarTreeDocuments(null, currentDocument, isMerge);
+                    while (_docId < numDocs) {
+                        StarTreeDocument doc;
+                        try {
+                            doc = getSegmentStarTreeDocument(sortedDocIds[_docId++], documentBytes, isMerge);
+                        } catch (IOException e) {
+                            throw new RuntimeException("Reducing documents failed ", e);
+                        }
+                        if (!Arrays.equals(doc.dimensions, next.dimensions)) {
+                            currentDocument = doc;
+                            return next;
+                        } else {
+                            next = reduceSegmentStarTreeDocuments(next, doc, isMerge);
+                        }
+                    }
+                    _hasNext = false;
+                    IOUtils.closeWhileHandlingException(finalSegmentDocsFileInput);
+                    try {
+                        tmpDirectory.deleteFile(segmentDocsFileOutput.getName());
+                    } catch (final IOException ignored) {}
+                    return next;
+                }
+            };
+        } catch (IOException ex) {
+            IOUtils.closeWhileHandlingException(segmentDocsFileInput);
+            throw ex;
+        }
+    }
+
+    /**
+     * Get segment star tree document from the segment.documents file
+     */
+    public StarTreeDocument getSegmentStarTreeDocument(int docID, long documentBytes, boolean isMerge) throws IOException {
+        return readStarTreeDocument(segmentRandomInput, docID * documentBytes, isMerge);
+    }
+
+    /**
+     * Get star tree document for the given docId from the star-tree documents file
+     */
+    @Override
+    public StarTreeDocument getStarTreeDocument(int docId) throws IOException {
+        ensureDocumentReadable(docId);
+        return readStarTreeDocument(starTreeDocsFileRandomInput, starTreeDocumentOffsets.get(docId), true);
+    }
+
+    // This should be only used for testing
+    @Override
+    public List<StarTreeDocument> getStarTreeDocuments() throws IOException {
+        List<StarTreeDocument> starTreeDocuments = new ArrayList<>();
+        for (int i = 0; i < numStarTreeDocs; i++) {
+            starTreeDocuments.add(getStarTreeDocument(i));
+        }
+        return starTreeDocuments;
+    }
+
+    @Override
+    public Long getDimensionValue(int docId, int dimensionId) throws IOException {
+        ensureDocumentReadable(docId);
+        return starTreeDocsFileRandomInput.readLong((starTreeDocumentOffsets.get(docId) + ((long) dimensionId * Long.BYTES)));
```

Comment:
Should this not handle null case?

Reply:
Null is never written to files. We always write 0 in place of null and use the bitset to read it back as null.